impl/memory/mutation_logs.go

```diff
@@ -119,7 +119,7 @@ func (m MutationLogs) HighWatermark(_ context.Context, _ string, logID int64, st
 	count := int32(0)
 	high := start // Preserve start time if there are no rows to process.
 	for ; i < len(logShard) && count < batchSize; i++ {
-		high = logShard[i].time.Add(time.Nanosecond) // Returns exclusive n + 1
+		high = logShard[i].time.Add(time.Microsecond) // Returns exclusive n + 1
```

Comment:
My first reaction to this line is "Why? The idea was that we just want the minimal watermark that's greater". Then I understood that you changed it to a microsecond because one of the storage implementations uses the microsecond granularity, and so +1 nanosecond would render "equal" because of the time-to-microsecond conversion error.

This is the 3rd reason to not use `time.Time`! Pleeeease. The idea why to use `time.Time` was that it would be independent from storage. The reality is that it's not. What if at some point another storage implementation wants a millisecond granularity? Would you convert all the code to add a millisecond instead of a microsecond? All we want is an abstract int, and do an abstract +1 here.

Reply:
The reason for this API leakage is perhaps an ugly hack buried inside of the Highwatermarks function -- Let's discuss that? 

Basically, Highwatermarks *should* return the highest primary key from the batch table, but because we've defined the ReadLog's API in terms of `[low, high)`,  we need to return a high watermark that's *just* beyond the primary key of the highest item.

Reply:
So, there are options:

1. `ReadLog` request is `[low, high)`. And then the `HighWatermarks` returns watermarks "just beyond" the last one, because these are going to be used as `low` in follow-up requests.
2. `ReadLog` request is in `(low, high]` range. `HighWatermarks` simply returns the last ID, and it can be used as a follow-up `low`. Cons: a) we can't `ReadLog` at position 0 in this case; b) the logs are sorted/read left-to-right, so `[low, high)` seems more intuitive.
3. Currently: `ReadLog` is in `[low, high)`, `HighWatermarks` returns the last ID, and the application layer does "logical" +1 (which tries to satisfy all existing storage implementations).
4. `ReadLog` is in `[low, high)`, `HighWatermarks` returns the last ID, and the application layer does "physical" +1 (independent of storage implementations; needs watermarks being ints or `Watermark struct` with a defined increment operation).

I think (1) is the preferred option, and if we choose it then the question of having `Watermark struct` is orthogonal (but still strongly voted for by me).

Reply:
I've come up with a way to go with option 1, and also stick with +1